manifests/addon.pp

```diff
@@ -1,72 +1,115 @@
 #
-# Alternatively you can feed inputs directly into splunk::addon using the
-# inputs parameter (useful if you are configuring from Hiera)
+# @param splunkbase_source
+#   When set the add-on will be installed from a splunkbase compatible archive instead of OS packages
 #
+# @param package_name
+#  The OS package to install if you are not installing via splunk compatible archive
 #
-# splunk::addon { 'search':
-#   package_manage => false,
-#   inputs          => {
-#     'monitor:///var/log/messages' => {
-#       'attributes' => {
-#         'index' => 'server_t',
-#       }
-#     }
-#   }
-# }
+# @param inputs
+#  A hash of inputs to be configured as part of add-on installation, alterntively you can also define splunk_input or splunkforwarder_input resouces seperately
 #
 define splunk::addon (
   Optional[Stdlib::Absolutepath] $splunk_home = undef,
   Boolean $package_manage                     = true,
+  Optional[String[1]] $splunkbase_source      = undef,
   Optional[String[1]] $package_name           = undef,
   Hash $inputs                                = {},
 ) {
 
-  include 'splunk::params'
+  include splunk::params
+
+  if defined(Class['splunk::forwarder']) {
+    $mode = 'forwarder'
+  } else {
+    $mode = 'enterprise'
+  }
 
   if $splunk_home {
     $_splunk_home = $splunk_home
   }
   else {
-    $_splunk_home = $splunk::params::forwarder_homedir
+    case $mode {
+      'forwarder':  { $_splunk_home = $splunk::params::forwarder_homedir }
+      'enterprise': { $_splunk_home = $splunk::params::enterprise_homedir }
+      default:      { fail('Instances of Splunk::Addon require the declaration of one of either Class[splunk::enterprise] or Class[splunk::forwarder]') }
+    }
   }
 
   if $package_manage {
-    package { $package_name:
-      ensure => installed,
-      before => File["${_splunk_home}/etc/apps/${name}/local"],
+    if $splunkbase_source {
+      $archive_name = $splunkbase_source.split('/')[-1]
+      archive { $name:
+        path         => "${splunk::params::staging_dir}/${archive_name}",
+        source       => $splunkbase_source,
+        extract      => true,
+        extract_path => "${_splunk_home}/etc/apps",
+        creates      => "${_splunk_home}/etc/apps/${name}",
+        cleanup      => true,
+        before       => File["${_splunk_home}/etc/apps/${name}/local"],
+      }
+    } else {
+      package { $package_name:
+        ensure => installed,
+        before => File["${_splunk_home}/etc/apps/${name}/local"],
+      }
     }
   }
 
-  file { "${_splunk_home}/etc/apps/${name}/local":
-    ensure => directory,
-  }
+  file { "${_splunk_home}/etc/apps/${name}/local": ensure => directory }
 
-  if $inputs {
-    concat { "splunk::addon::inputs_${name}":
-      path    => "${_splunk_home}/etc/apps/${name}/local/inputs.conf",
-      require => File["${_splunk_home}/etc/apps/${name}/local"],
+  unless $inputs.empty {
```

Comment:
Why? Iterating an empty hash is fine.

Reply:
You are correct. Just didn't realize. Tested just now and works fine without testing first